### PR TITLE
Timer: Adds support for pomodoro timer

### LINKED ISSUE #4186: Timer: Trigger on "Pomodoro timer"
I think it would be a good idea to make the search 'pomodoro timer' trigger a timer of 15 minutes.
If you didn't know: [pomodoro technique](https://en.wikipedia.org/wiki/Pomodoro_Technique)

------
IA Page:  http://duck.co/ia/view/timer 
[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @mattr555

### PR DESCRIPTION
**Pull request by adityatandon007**
## Description of new Instant Answer, or changes
This PR adds the support for pomodoro timer

## Related Issues and Discussions

Fixes #4186 

## People to notify
@pjhampton @mintsoft 

## Testing & Review
To be completed by Community Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [ ] Title follows correct format (Specifies Instant Answer + Purpose)
- [ ] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [ ] Instant Answer page is correctly filled out and contains:
    - The **Title** is appropriately named and formatted
    - The **IA topics** are present and appropriate
    - The **Description** is clear and coherent
    - **Source Name** exists if applicable
    - All **Example Queries** trigger on [Beta](https://beta.duckduckgo.com/)

**Code**
- [ ] Adheres to the [DuckDuckGo Style Guide](https://docs.duckduckhack.com/resources/code-style-guide.html)
- [ ] Behaviour is appropriately tested. If improvement, tests are adequately extended.
- [ ] There is no unnecessary files in place (such as editor config files)
- [ ] There is no API keys / secrets present
- [ ] Tests are passing (run `$ duckpan test <goodie_id>`)
    - Tester should report any failures

**Ready to merge?**

- [ ] Has this IA been deployed to and tested on [beta.duckduckgo.com](https://beta.duckduckgo.com/)?
- [ ] For larger commits, has this been approved be more than one community member?
- [ ] The number of reviews is appropriate for this type of PR
- [ ] The commit message is clear, coherent and fitting

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/timer

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

**Comment by adityatandon007** [comment]
@pjhampton Though I have added the support for pomodoro timer, I couldn't find a hack to default it to `25 minutes`. For now it is defaulting to `1 min` like the rest of the triggers. Let me know if you find a solution to default it to `25 min` :slightly_smiling_face:

**Comment by adityatandon007** [comment]
@pjhampton One more thing I haven't added that `pomodoro clock` thing. As far as I know clock doesn't mean the same as `timer`. Clock is a device like a watch which displays/indicates time. So, if somebody says clock they mean that device not the `timer`. If you want `pomodoro clock` to be added then I think we should also add `pomodoro watch` and similarly someone would say to add `pomodoro pendulum` and someone would even go on saying to add `pomodoro smart watch`. 

I think every IA should be very specific with the result it serve.
**Timer is timer and Clock is clock**

File: t/Timer.t [hunk 1]
```diff
@@ -102,6 +105,10 @@ ddg_goodie_test(
     '20 mins'                                        => build_test('1200'),
     'timer 20 minutes'                               => build_test('1200'),
     'timer 60 minutes'                               => build_test('3600'),
+				'pomodoro timer for 10 mins'                     => build_test('600'),
```
Comment (moollaza) [inline t/Timer.t]: @adityatandon007 please fix the indentation

**Comment by adityatandon007** [comment]
@moollaza done indentation!

**Comment by adityatandon007** [comment]
@moollaza One more thing how can i default it to 25 minutes. Any suggestions would be great.

**Comment by mattr555** [comment]
@adityatandon007 You can edit the `parse_query_for_time` function. Here's what I came up with:

```perl
sub parse_query_for_time {
    my $query = shift;
    my $orig_query = $query;
    # <snip>
    return 25*60 if ($time == 0 && $orig_query =~ /pomodoro/);
    return ($time <= $MAX_TIME) ? $time : $MAX_TIME;
}
```
Please also edit your tests when you add this, thanks! :smile:

**Comment by adityatandon007** [comment]
@mattr555 thanks a lot , but there is a slight awkwardness when I type `pomodoro timer` it starts `automatically` whereas on other cases where the default is set to `1 minute` you got a `play` button to choose when you want it to start. Is there a workaround so it doesn't start automatically.

**Comment by mintsoft** [comment]
@adityatandon007 This has conflicts now, can you rebase?

**Comment by adityatandon007** [comment]
@mintsoft I have resolved the conflicts but pomodoro timer starts automatically without letting the user press the play button. Do you have something to overcome this

File: lib/DDG/Goodie/Timer.pm [hunk 2]
```diff
@@ -63,6 +65,7 @@ sub parse_query_for_time {
         }
         $query =~ s/$timer_re//;
     }
+		return 25*60 if ($time == 0 && $orig_query =~ /pomodoro/);
```
Comment (mintsoft) [inline lib/DDG/Goodie/Timer.pm]: Can you replace this hard-tab with spaces?

**Comment by mintsoft** [comment]
@adityatandon007 You'll basically have to extend the behaviour of https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/share/goodie/timer/timer.js#L424 here you could pass in the `structured_answer` something should `should_autostart` which is 0 if pomodoro is used to trigger the query and 1 otherwise?

**Comment by adityatandon007** [comment]
@mintsoft I have added the functionality of the play button, which is now appearing for pomodoro queries but I am not able to solve the test cases but the duckpan server is working fine for those queries. I think I didn't know how to modify the test case, can you help me on how to resolve this.

File: t/Timer.t [hunk 3]
```diff
@@ -14,7 +14,9 @@ my $goodieVersion = $DDG::GoodieBundle::OpenSourceDuckDuckGo::VERSION // 999;
 
 sub build_structured_answer {
     my $time = shift;
-    $time = $time || 0;
+    my $autoPlay = shift;
+    $autoPlay = $autoPlay || 0;
```
Comment (mintsoft) [inline t/Timer.t]: I'd remove the defaulting to `0`here and on `$time` so it's clearer what's going on

File: t/Timer.t [hunk 4]
```diff
@@ -55,57 +58,68 @@ ddg_goodie_test(
...
-    'start countdown timer for 77 mins 13 secs'      => build_test('4633'),
-    '10 minutes'                                     => build_test('600'),
-    '20 mins'                                        => build_test('1200'),
-    'timer 20 minutes'                               => build_test('1200'),
-    'timer 60 minutes'                               => build_test('3600'),
-    'start a timer for 20 minutes'                   => build_test('1200'),
-    'set a timer to 60 minutes'                      => build_test('3600'),
-    'set a timer of 60 minutes'                      => build_test('3600'),
-    'timer of 5:15 mins'                             => build_test('315'),          
+    'timer 15 mins'                                  => build_test('900',1),
```
Comment (mintsoft) [inline t/Timer.t]: So all of these tests you've got `,1` which in the `build_sub` you've made mean autostart, so for the tests that are failing this `1` should be `0`

**Comment by adityatandon007** [comment]
@mintsoft Thanks for the help. Updated the test and all exited with status `successful` . Let me know in case you need any other changes

**Comment by mintsoft** [comment]
@moollaza this LGTM

**Comment by pjhampton** [comment]
We discussed this PR internally and have decided to not merge it for the time being. Personally, I feel this solution is overkill for the problem it is trying to solve. We can come back and look at this again someday when DDH exits maintenance mode.

**Pull request closed without merging**